File: numberTheory/prime.py
# problem looking at solving for primes using a few methods and testing the time
from __future__ import division
from sys import argv
from math import *
from matplotlib.patches import Patch
from pylab import *
from datetime import datetime
import subprocess


script, prime_number = argv


t0 = datetime.now()
data= subprocess.Popen('./prime %s' % prime_number, shell=True, stdout=subprocess.PIPE,stderr=subprocess.PIPE,stdin=subprocess.PIPE).communicate()
data = data[0].split(",")

print "Process | Time(s) | Standard Deviation | Max Error" 

x_values = data[data.index('X Values Start')+1:data.index("X Values End")]
for i in range(0,len(x_values)):
   x_values[i] = float(x_values[i])
t1 = datetime.now()
time = (t1-t0).seconds
#print "%s | %s | %.3f | %.3f " % ("Upwind", time, std, max_element)

y_values = data[data.index('Y Values Start')+1:data.index("Y Values End")]
for i in range(0,len(y_values)):
   y_values[i] = float(y_values[i])

x_lagrange = data[data.index('X Lagrange Start')+1:data.index("X Lagrange End")]
for i in range(0,len(x_lagrange)):
   x_lagrange[i] = float(x_lagrange[i])

y_lagrange = data[data.index('Y Lagrange Start')+1:data.index("Y Lagrange End")]
for i in range(0,len(y_lagrange)):
   y_lagrange[i] = float(y_lagrange[i])

x_lagrange_values = data[data.index('X Lagrange Values Start')+1:data.index("X Lagrange Values End")]
for i in range(0,len(x_lagrange_values)):
   x_lagrange_values[i] = float(x_lagrange_values[i])

y_lagrange_values = data[data.index('Y Lagrange Values Start')+1:data.index("Y Lagrange Values End")]
for i in range(0,len(y_lagrange_values)):
   y_lagrange_values[i] = float(y_lagrange_values[i])





plt.plot(x_values, y_values, mfc='red', ms=12, label='Analytical')
plt.plot(x_lagrange_values, y_lagrange_values,'x')
plt.plot(x_lagrange, y_lagrange)
plt.xlabel('X Values')
plt.ylabel('Y Values')
plt.title('First Plot')
plt.axis([1,15,-1.5,1.5])
legend()
show()

File: numberTheory/prime.cpp
#include <iostream>
#include <string>
#include <cstdlib> 
#include <cmath>


using namespace std;

void printarray (double arg[], int length, string input) {
  for (int i=0; i<length; i++)
      if (i==0){
         cout << input <<" Start" << "," << arg[i] << ",";
      }else if (i==length-1){
         cout << arg[i] << "," << input <<" End" << ",";
      }else{
         cout << arg[i] << ",";
      }
}

double func(double x, int prime){
	double PI = 3.141592654;
	return cos(2*PI*x)*cos(2*PI*prime/x);

}

double lagrange_interpolation(double x, double x_values[], double y_values[], int degree){
	double L = 0;
	double f_at_x;
	for(int i=0; i<= degree; i++){
		f_at_x = y_values[i];
		for(int j=0; j<= degree; j++){
			if (i != j){
				f_at_x  = f_at_x * (x-x_values[j])/(x_values[i] - x_values[j]);
			}
		}
		L = L + f_at_x;
	}
	return L;
}




int main(int argc, char **argv){

	int x_spaces = 100;
	int degree = 25;
   int prime = atof(argv[1]);
	int x_start = 1;
	int x_end = prime;
	double x_values[x_spaces+1];
	double y_values[x_spaces+1];
	double x_lagrange[x_spaces+1];
	double y_lagrange[x_spaces+1];
	double x_lagrange_values[degree+1];
	double y_lagrange_values[degree+1];

	// This is for the analytical answer
	for(int i=0;i<x_spaces;i++){
		x_values[i] = x_start + (x_end-x_start) / (1.0*x_spaces-1.0) *(i);
		y_values[i] = func(x_values[i], prime);
	}
	
	// the points to be used for Lagrangian
	for(int i=0;i<degree+1;i++){
		x_lagrange_values[i] = x_start + (x_end-x_start) / (1.0*degree) *(i);
		y_lagrange_values[i] = func(x_lagrange_values[i],prime);
	}

	for(int i=0;i<x_spaces;i++){
		x_lagrange[i] = x_values[i];
		y_lagrange[i] = lagrange_interpolation(x_lagrange[i],x_lagrange_values,y_lagrange_values,degree);
	}



	printarray(x_values, x_spaces, "X Values");
	printarray(y_values, x_spaces, "Y Values");
	printarray(x_lagrange_values, degree+1, "X Lagrange Values");
	printarray(y_lagrange_values, degree+1, "Y Lagrange Values");
	printarray(x_lagrange, x_spaces, "X Lagrange");
	printarray(y_lagrange, x_spaces, "Y Lagrange");


   return 0; 


}

